src/spi_comms.c

```diff
@@ -0,0 +1,357 @@
+    opt_set_data_rdy_high();
+}
+
+
+// return status of SS pin
+uint8_t opt_check_ss_pin(void){
+    return SS_PIN & _BV(SS); 
+}
+
+
+// set DATA_RDYn low
+void opt_set_data_rdy_low(){
+    DATA_RDYn_PORT &= ~_BV(DATA_RDYn);     // set output LOW
+}
+
+// set DATA_RDYn high
+void opt_set_data_rdy_high(){
+    DATA_RDYn_PORT |= _BV(DATA_RDYn);  // set output HIGH
+}
+
+
+// to be put in infinite loop in main
+void opt_loop(void){
+    // if SPI transfer if completed
+    if (SPSR & _BV(SPIF)){
+        spi_first_byte = SPDR;
+        opt_set_data_rdy_low();         // set DATA_RDYn low to signal to PAY that it's ready for the second byte
+
+        // wait until another SPI transfer is completed
+        // --> aka wait until SPIF is no longer high 
+        while (!(SPSR & _BV(SPIF)));
+
+        opt_set_data_rdy_high();        // clear DATA_RDYn, so PAY doesn't freak out
+        spi_second_byte = SPDR;
+
+        // now, got both bytes
+        // perform the requested command and send back data if necessary
+        manage_cmd(spi_first_byte);
+    }
+    opt_set_data_rdy_high();
+}
+
+
+// depending on cmd_code, does appropriate requested function + return data (if needed)
+void manage_cmd (uint8_t cmd_code){
+    // if first byte is get_reading, then 2nd byte is well data
+    if (cmd_code == CMD_GET_READING){
+        // spi_second_byte contains well_data
+        opt_update_reading(spi_second_byte);    // performs reading (3 bytes), stores it in wells[32] of well_t
+        opt_transfer_reading();       // shifts reading data into SPDR over 3 SPI transmissions
+    }
+
+    // invalid command
+    else{ 
+        spi_status_byte |= _BV(SPI_ERROR_BIT) | _BV(SPI_INVALID_COMMAND_BIT);
+        SPDR = spi_status_byte;
+        opt_set_data_rdy_low();
+
+        // wait until SPI transfer is complete
+        while (!(SPSR & _BV(SPIF)));
+        opt_set_data_rdy_high();
+    }
+
+}
+
+
+// calibrate and take well readings
+// well_data[7] - optical density = 0, fluorescent LED = 1
+// well_data[5:0] - well number (0-31)
+void opt_update_reading(uint8_t well_data){
+    update_well_reading((well_data && 0x1F), well_data >> 7);
+}
+
+
+void opt_transfer_reading(){
+    uint32_t reading = 0;
+    if (spi_second_byte >> 7 == PAY_OPTICAL)    // bit 7 = 0
+        reading = (wells + (spi_second_byte && 0x1F))->last_opt_reading;
+    else // PAY_LED, bit 7 = 1
+        reading = (wells + (spi_second_byte && 0x1F))->last_led_reading;
+
+    uint8_t shift = 16;
+    while (shift >= 0){
+        // load the next byte of data, ready for SPI transmission out
+        spi_spdr_tx_buffer = (uint8_t)(reading >> shift);
+        opt_set_data_rdy_low();     // signal to PAY to initiate SPI transfer
+
+        // wait until SPI transfer is complete
+        while (!(SPSR & _BV(SPIF)));
+        opt_set_data_rdy_high();
+
+        shift = shift - 8;
+
+        // might need small delay here, will test

```

Comment:
This will always infinite loop because `shift` is unsigned so `shift >= 0` is always true because subtracting 8 will either equal 0 or underflow back to 0xFF.